Xcode is not able to attach to WebContent Development variant

pvollan opened the pull request:
#### c663098fb4c10a2d32d9879df7fdec7606ed3edd
<pre>
Xcode is not able to attach to WebContent Development variant
<a href="https://bugs.webkit.org/show_bug.cgi?id=274676">https://bugs.webkit.org/show_bug.cgi?id=274676</a>
<a href="https://rdar.apple.com/128695932">rdar://128695932</a>

Reviewed by Chris Dumez.

Restricted entitlement should not be added to WebContent Development variant. Xcode will not be
able to attach to the Development version if there are restricted entitlements added to it.

* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):

Canonical link: <a href="https://commits.webkit.org/279416@main">https://commits.webkit.org/279416@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c60f5cd14c4150196448c47ae54935bcbf82709

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53380 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32732 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5879 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56660 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4106 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40184 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3871 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43281 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2692 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55478 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30947 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46115 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24413 "Found 2 new API test failures: /WPE/TestMultiprocess:/webkit/WebKitWebView/multiprocess-create-ready-close, /WPE/TestWebProcessExtensions:/webkit/WebKitWebView/web-process-crashed (failure)") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/53224 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27814 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3435 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2262 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46740 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49575 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3607 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58255 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52894 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28531 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3600 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50680 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29744 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46317 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50010 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30668 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/65200 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7859 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29511 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12429 "Found 244 jsc stress test failures: basic-tests.yaml/stress-test.js.bytecode-cache, basic-tests.yaml/stress-test.js.dfg-eager, basic-tests.yaml/stress-test.js.dfg-eager-no-cjit-validate, basic-tests.yaml/stress-test.js.eager-jettison-no-cjit, basic-tests.yaml/stress-test.js.no-cjit-collect-continuously ...") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -1572,6 +1572,9 @@ static float currentBacklightLevel()
 #if ENABLE(NOTIFY_BLOCKING)
 void WebProcess::postNotification(const String& message, std::optional<uint64_t> state)
 {
+    if (!WTF::processHasEntitlement("com.apple.developer.web-browser-engine.restrict.notifyd"_s))
```
I don't get it. Why is it OK to ignore notifications when the entitlement is missing? Shouldn't we disable notification forwarding when the entitlement is missing? So that this code path is not even exercised?

pvollan replied:
Ah, good point! I moved the check to the UI process instead in the last version of the patch. The last patch is still registering for the notifications, since we do not know whether the WebContent processes will have this entitlement at this point in time.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -765,7 +765,11 @@ static void logProcessPoolState(const WebProcessPool& pool)
                 String messageString(message);
-                protectedThis->sendToAllProcesses(Messages::WebProcess::PostNotification(messageString, (status == NOTIFY_STATUS_OK) ? std::optional<uint64_t>(state) : std::nullopt));
+                for (auto& process : protectedThis->m_processes) {
+                    if (process->auditToken() && !WTF::hasEntitlement(process->auditToken().value(), "com.apple.developer.web-browser-engine.restrict.notifyd"_s))
```
Are processes without this entitlement allowed to listen to notifications in-process then? Seems like they should. Yet, I don't see any sandbox changes in this PR. I find it a little surprising this code is even running when the processes don't have notifyd hardening.

pvollan replied:
Ah, that is a good point! Yes, processes without this entitlement are allowed to listen to notifications, because the sandbox rule is using the presence of the entitlement to determine if the notification service should be blocked.

pvollan replied:
Also, as you point out, ideally we should not be observing these notifications in the UI process when launching the Development version of the WebContent process. However, at the time we register for the notifications, we have not launched a WebContent process yet and do not know if the entitlement will be present. In the future, we may also want to forward the notifications to the Networking and GPU process for hardening purposes, and we would then need to register for the notifications, regardless of whether the Development version of WebContent is being used or not.

pvollan committed:
Xcode is not able to attach to WebContent Development variant
https://bugs.webkit.org/show_bug.cgi?id=274676
rdar://128695932

Reviewed by Chris Dumez.

Restricted entitlement should not be added to WebContent Development variant. Xcode will not be
able to attach to the Development version if there are restricted entitlements added to it.

* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):

Canonical link: https://commits.webkit.org/279416@main
